commands/fetch_templates.go

```diff
@@ -18,113 +19,161 @@ import (
 	"github.com/openfaas/faas-cli/proxy"
 )
 
-const ZipFileName string = "master.zip"
+const (
+	defaultTemplateRepository = "https://github.com/openfaas/faas-cli"
+	templateDirectory         = "./template/"
+)
 
 // fetchTemplates fetch code templates from GitHub master zip file.
-func fetchTemplates(templateUrl string) error {
+func fetchTemplates(templateURL string, overwrite bool) error {
+	var existingLanguages []string
+	availableLanguages := make(map[string]bool)
+	var fetchedTemplates []string
+
+	if len(templateURL) == 0 {
+		templateURL = defaultTemplateRepository
+	}
 
-	err := fetchMasterZip(templateUrl)
+	archive, err := fetchMasterZip(templateURL)
+	if err != nil {
+		removeArchive(archive)
+		return err
+	}
 
-	zipFile, err := zip.OpenReader(ZipFileName)
+	zipFile, err := zip.OpenReader(archive)
 	if err != nil {
 		return err
 	}
 
-	log.Printf("Attempting to expand templates from %s\n", ZipFileName)
+	log.Printf("Attempting to expand templates from %s\n", archive)
 
 	for _, z := range zipFile.File {
-		relativePath := strings.Replace(z.Name, "faas-cli-master/", "", -1)
-		if strings.Index(relativePath, "template") == 0 {
-			fmt.Printf("Found \"%s\"\n", relativePath)
-			rc, err := z.Open()
-			if err != nil {
-				return err
-			}
+		var rc io.ReadCloser
 
-			err = createPath(relativePath, z.Mode())
-			if err != nil {
-				return err
-			}
+		relativePath := z.Name[strings.Index(z.Name, "/")+1:]
+		if strings.Index(relativePath, "template/") != 0 {
+			// Process only directories inside "template" at root
+			continue
+		}
+
+		var language string
+		if languageSplit := strings.Split(relativePath, "/"); len(languageSplit) > 2 {
+			language = languageSplit[1]
+
+			if len(languageSplit) == 3 && relativePath[len(relativePath)-1:] == "/" {
```

Comment:
Pre-squash I asked for these magic numbers to be documented / split out either into a new function or named consts. If I had to work on this I wouldn't know what to do.